File: app/src/main/java/net/squanchy/schedule/view/TalkEventItemView.java
package net.squanchy.schedule.view;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.TextView;

import net.squanchy.R;
import net.squanchy.eventdetails.widget.ExperienceLevelView;
import net.squanchy.schedule.domain.view.Event;
import net.squanchy.support.lang.Optional;
import net.squanchy.support.widget.SpeakerView;

import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

public class TalkEventItemView extends EventItemView {

    private TextView titleView;
    private TextView timestampView;
    private ExperienceLevelView experienceLevelView;
    private SpeakerView speakerView;

    public TalkEventItemView(Context context, AttributeSet attrs) {
        this(context, attrs, R.attr.cardViewDefaultStyle);
    }

    public TalkEventItemView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        timestampView = (TextView) findViewById(R.id.timestamp);
        titleView = (TextView) findViewById(R.id.title);
        experienceLevelView = (ExperienceLevelView) findViewById(R.id.experience_level);
        speakerView = (SpeakerView) findViewById(R.id.speaker_container);
    }

    @Override
    public void updateWith(Event event) {
        ensureSupportedType(event.getType());

        timestampView.setText(startTimeAsFormattedString(event));
        titleView.setText(event.getTitle());

        if (event.getExperienceLevel().isPresent()) {
            experienceLevelView.setExperienceLevel(event.getExperienceLevel().get());
            experienceLevelView.setVisibility(VISIBLE);
        } else {
            experienceLevelView.setVisibility(INVISIBLE);
        }

        speakerView.setVisibility(event.getSpeakers().isEmpty() ? GONE : VISIBLE);
        speakerView.updateWith(event.getSpeakers(), Optional.absent());
    }

    private String startTimeAsFormattedString(Event event) {
        DateTimeFormatter formatter = DateTimeFormat.shortTime()
                .withZone(event.getTimeZone());

        return formatter.print(event.getStartTime().toDateTime());
    }

    private void ensureSupportedType(Event.Type type) {
        if (type == Event.Type.TALK || type == Event.Type.KEYNOTE) {
            return;
        }
        throw new IllegalArgumentException("Event with type " + type.name() + " is not supported by this view");
    }
}

File: app/src/main/java/net/squanchy/schedule/view/TalkEventItemView.kt
package net.squanchy.schedule.view

import android.content.Context
import android.util.AttributeSet
import android.view.View
import androidx.core.view.isVisible
import kotlinx.android.synthetic.main.item_schedule_event_talk.view.*
import net.squanchy.R
import net.squanchy.schedule.domain.view.Event
import net.squanchy.schedule.domain.view.Place
import net.squanchy.support.content.res.getColorFromAttribute
import net.squanchy.support.graphics.toColorStateList
import net.squanchy.support.lang.getOrThrow
import org.joda.time.format.DateTimeFormat
import java.util.Locale

class TalkEventItemView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet,
    defStyleAttr: Int = R.attr.cardViewDefaultStyle
) : EventItemView(context, attrs, defStyleAttr) {

    private val timeFormatter = DateTimeFormat.shortTime()
    private val dateFormatter = DateTimeFormat.forPattern("EEE d")

    override fun updateWith(event: Event, showRoom: Boolean, showDay: Boolean, showFavorite: Boolean) {
        ensureSupportedType(event.type)

        timestamp.text = startTimeAsFormattedString(event, showDay)
        title.text = event.title
        room.apply {
            text = event.place.orNull().toPlaceLabel()
            isVisible = showRoom
        }

        when {
            event.type == Event.Type.KEYNOTE -> {
                experienceLevel.text = event.type.name.toLowerCase(Locale.getDefault())
                experienceLevel.backgroundTintList = context.theme.getColorFromAttribute(R.attr.colorAccent).toColorStateList()
                experienceLevel.isVisible = true
            }
            event.experienceLevel.isDefined() -> {
                experienceLevel.setExperienceLevel(event.experienceLevel.getOrThrow())
                experienceLevel.isVisible = true
            }
            else -> experienceLevel.isVisible = false
        }

        speakersContainer.visibility = if (event.speakers.isEmpty()) View.GONE else View.VISIBLE
        speakersContainer.updateWith(event.speakers, null)

        favoriteIcon.isVisible = if (showFavorite) event.favorite else false
    }

    private fun Place?.toPlaceLabel(): CharSequence? = if (this != null) " • ${this.name}" else null

    private fun ensureSupportedType(type: Event.Type) {
        when (type) {
            Event.Type.TALK, Event.Type.KEYNOTE, Event.Type.WORKSHOP -> return
            else -> throw IllegalArgumentException("Event with type ${type.name} is not supported by this view")
        }
    }

    private fun startTimeAsFormattedString(event: Event, showDay: Boolean): String {
        val timeZone = event.timeZone
        val startDateTime = event.startTime.toDateTime(timeZone)

        // Note: the space at the end of the template string is not a typo, it separates the date from the time
        val formattedDate = if (showDay) "${dateFormatter.withZone(timeZone).print(startDateTime)} " else ""
        val formattedTime = timeFormatter.withZone(timeZone).print(startDateTime)

        return "$formattedDate$formattedTime"
    }
}
